tests/rustc_tests.rs

```diff
@@ -0,0 +1,761 @@
+            .fold(true)
+            .annotation(Level::Error.span(17..27).label("`X` is a good letter"))
+            .annotation(Level::Error.span(31..55).label("`Y` is a good letter too")),
+    );
+
+    // This is all `^` until we support primary and secondary labels
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:6
+  |
+3 |      X0 Y0 Z0
+  |  _______^
+4 | |    X1 Y1 Z1
+  | | ____^____^
+  | ||____|
+  |  |    `X` is a good letter
+5 |  |   X2 Y2 Z2
+6 |  |   X3 Y3 Z3
+  |  |__________^ `Y` is a good letter too
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
+#[test]
+fn multiple_labels_primary_without_message() {
+    let source = r#"
+fn foo() {
+  a { b { c } d }
+}
+"#;
+    let input = Level::Error.title("foo").snippet(
+        Snippet::source(source)
+            .line_start(1)
+            .origin("test.rs")
+            .fold(true)
+            .annotation(Level::Error.span(18..25).label(""))
+            .annotation(Level::Error.span(14..27).label("`a` is a good letter"))
+            .annotation(Level::Error.span(22..23).label("")),
+    );
+
+    // This is all `^` until we support primary and secondary labels
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:7
+  |
+3 |   a { b { c } d }
+  |   ^^^^^^^^^^^^^ `a` is a good letter
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
+#[test]
+fn multiple_labels_secondary_without_message() {
+    let source = r#"
+fn foo() {
+  a { b { c } d }
+}
+"#;
+    let input = Level::Error.title("foo").snippet(
+        Snippet::source(source)
+            .line_start(1)
+            .origin("test.rs")
+            .fold(true)
+            .annotation(Level::Error.span(14..27).label("`a` is a good letter"))
+            .annotation(Level::Error.span(18..25).label("")),
+    );
+
+    // This is all `^` until we support primary and secondary labels
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:3
+  |
+3 |   a { b { c } d }
+  |   ^^^^^^^^^^^^^ `a` is a good letter
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
+#[test]
+fn multiple_labels_primary_without_message_2() {
+    let source = r#"
+fn foo() {
+  a { b { c } d }
+}
+"#;
+    let input = Level::Error.title("foo").snippet(
+        Snippet::source(source)
+            .line_start(1)
+            .origin("test.rs")
+            .fold(true)
+            .annotation(Level::Error.span(18..25).label("`b` is a good letter"))
+            .annotation(Level::Error.span(14..27).label(""))
+            .annotation(Level::Error.span(22..23).label("")),
+    );
+
+    // This is all `^` until we support primary and secondary labels
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:7
+  |
+3 |   a { b { c } d }
+  |   ^^^^^^^^^^^^^
+  |       |
+  |       `b` is a good letter
+  |
+"#]];
```

Comment:
For this case, you really want to have different label levels, so that you can confirm that overlapping spans of different levels do not "hide" each other. In `rustc` we ensure we print from widest to shortest span, so that this ends up like `^^^^-----^^^^` or `----^^^^----`. This isn't *perfect* but it is the best we can do when restricted to ASCII and a single line of text.

Reply:
... yeah, I think my comment from the other issue stands.  I think rustc is being too "smart" and a lot of these overlapping spans and weird cases are bonkers for a user to decipher